mesa_frames/abstract/space.py

```diff
@@ -733,3 +738,419 @@ def full_cells(self) -> DataFrame:
         return self._sample_cells(
             None, with_replacement=False, condition=self._full_cell_condition
         )
+
+
+class GridDF(DiscreteSpaceDF):
+    """The GridDF class is an abstract class that defines the interface for all grid classes in mesa-frames.
+    Inherits from DiscreteSpaceDF.
+
+    Methods
+    -------
+    __init__(model: 'ModelDF', dimensions: Sequence[int], torus: bool = False, capacity: int | None = None, neighborhood_type: str = 'moore')
+        Create a new GridDF object.
+    out_of_bounds(pos: GridCoordinate | GridCoordinates) -> DataFrame
+        Check whether the input positions are out of bounds in a non-toroidal grid.
+
+    Properties
+    ----------
+    dimensions : Sequence[int]
+        The dimensions of the grid
+    neighborhood_type : Literal['moore', 'von_neumann', 'hexagonal']
+        The type of neighborhood to consider
+    torus : bool
+        If the grid is a torus
+    """
+
+    _grid_capacity: (
+        GridCapacity  # Storing the remaining capacity of the cells in the grid
+    )
+    _neighborhood_type: Literal[
+        "moore", "von_neumann", "hexagonal"
+    ]  # The type of neighborhood to consider
+    _offsets: DataFrame  # The offsets to compute the neighborhood of a cell
+    _torus: bool  # If the grid is a torus
+
+    def __init__(
+        self,
+        model: "ModelDF",
+        dimensions: Sequence[int],
+        torus: bool = False,
+        capacity: int | None = None,
+        neighborhood_type: str = "moore",
+    ):
+        """Grid cells are indexed, where [0, ..., 0] is assumed to be the
+        bottom-left and [dimensions[0]-1, ..., dimensions[n]-1] is the top-right. If a grid is
+        toroidal, the top and bottom, and left and right, edges wrap to each other.
+
+        Parameters
+        ----------
+        model : 'ModelDF'
+            The model selfect to which the grid belongs
+        dimensions: Sequence[int]
+            The dimensions of the grid
+        torus : bool, optional
+            If the grid should be a torus, by default False
+        capacity : int | None, optional
```

Comment:
Is the `optional` at the end the usual convention for pydocstring? I thought `int | None` already means `Optional[int]`.

Reply:
I believe this is the standard behavior by [autoDocstring](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) , because `optional` only means that it has a default value, not that it's `None`.
But I will try with pydocstring.

Reply:
Look here (https://github.com/NilsJPWerner/autoDocstring/blob/HEAD/docs/numpy.md), optional means that the parameter has a default value..

Reply:
OK sounds good.